src/Zipper.ts

```diff
@@ -0,0 +1,279 @@
+import { Semigroup } from './Semigroup'
+import { Monoid } from './Monoid'
+import { Option, none, some } from './Option'
+import { isOutOfBound, snoc, take, drop, array, isEmpty, cons } from './Array'
+import { Applicative1, Applicative } from './Applicative'
+import { NonEmptyArray } from './NonEmptyArray'
+import { Foldable1 } from './Foldable'
+import { HKT } from './HKT'
+import { Traversable1 } from './Traversable'
+import { Comonad1 } from './Comonad'
+import { toString } from './function'
+
+/*
+  Adapted from
+
+  - https://github.com/DavidHarrison/purescript-list-zipper
+  - https://github.com/thunklife/purescript-zipper
+  - https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/Zipper.scala
+*/
+
+declare module './HKT' {
+  interface URI2HKT<A> {
+    Zipper: Zipper<A>
+  }
+}
+
+export const URI = 'Zipper'
+
+export type URI = typeof URI
+
+/**
+ * Provides a pointed array, which is a non-empty zipper-like array structure that tracks an index (focus)
+ * position in an array. Focus can be moved forward and backwards through the array.
+ *
+ * @data
+ * @constructor Zipper
+ * @since 1.9.0
+ * @example
+ *
+ * The array `[1, 2, 3, 4]` with focus on `3` is represented by `new Zipper([1, 2], 3, [4])`
+ */
+export class Zipper<A> {
+  readonly _A!: A
+  readonly _URI!: URI
+  length: number
+  constructor(readonly lefts: Array<A>, readonly focus: A, readonly rights: Array<A>) {
+    this.length = lefts.length + 1 + rights.length
+  }
+  /**
+   * Update the focus in this zipper.
+   * @since 1.9.0
+   */
+  update(a: A): Zipper<A> {
+    return new Zipper(this.lefts, a, this.rights)
+  }
+  /**
+   * Apply `f` to the focus and update with the result.
+   * @since 1.9.0
+   */
+  modify(f: (a: A) => A): Zipper<A> {
+    return this.update(f(this.focus))
+  }
+  /**
+   * @since 1.9.0
+   */
+  toArray(): Array<A> {
+    return snoc(this.lefts, this.focus).concat(this.rights)
```

Comment:
Just wondering if this.lefts.concat(cons(this.focus, this.rights)) would be more efficient.

This assumes cons to be constant time, and snoc being linear.

Reply:
ah no, I see cons is linear too, so no difference there